src/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -453,37 +465,73 @@ private int InternalValueCountCore()
-            int namelen;
+            List<string> names = new List<string>(values);
+
+            // Names in the registry aren't usually very long, although they can go to as large
+            // as 16383 characters (MaxValueLength).
+            //
+            // Every call to RegEnumValue will allocate another buffer to get the data from
+            // NtEnumerateValueKey before copying it back out to our passed in buffer. This can
+            // add up quickly- we'll try to keep the memory pressure low and grow the buffer
+            // only if needed.
 
-            fixed (char* namePtr = &name[0])
+            char[] name = ArrayPool<char>.Shared.Rent(100);
+
+            try
             {
-                for (int i = 0; i < values; i++)
+                int index = 0;
+                int result;
+                int nameLength = name.Length;
+
+                while ((result = Interop.Advapi32.RegEnumValue(
+                    _hkey,
+                    index,
+                    name,
+                    ref nameLength,
+                    IntPtr.Zero,
+                    null,
+                    null,
+                    null)) != Interop.Errors.ERROR_NO_MORE_ITEMS)
                 {
-                    namelen = name.Length;
-
-                    int ret = Interop.Advapi32.RegEnumValue(_hkey,
-                        i,
-                        namePtr,
-                        ref namelen,
-                        IntPtr.Zero,
-                        null,
-                        null,
-                        null);
-
-                    if (ret != 0)
+                    switch (result)
                     {
-                        // ignore ERROR_MORE_DATA if we're querying HKEY_PERFORMANCE_DATA
-                        if (!(IsPerfDataKey() && ret == Interop.Errors.ERROR_MORE_DATA))
-                            Win32Error(ret, null);
+                        // The size is only ever reported back correctly in the case
+                        // of ERROR_SUCCESS. It will almost always be changed, however.
+                        case Interop.Errors.ERROR_SUCCESS:
+                            names.Add(new string(name, 0, nameLength));
+                            index++;
+                            break;
+                        case Interop.Errors.ERROR_MORE_DATA:
+                            if (IsPerfDataKey())
+                            {
+                                // Enumerating the values for Perf keys always returns ERROR_MORE_DATA.
+                                names.Add(new string(name));
+                                index++;
+                                break;
+                            }
+                            else
+                            {
+                                nameLength = name.Length;
+                                ArrayPool<char>.Shared.Return(name);
+                                name = ArrayPool<char>.Shared.Rent(nameLength + 100);
```

Comment:
Why `+100` rather than `* 2`?  The latter is more common, and with our ArrayPool implementation, it's going to effectively do the latter anyway. `+100` just seems arbitrary.